File: JsonSchemaAndStructs/KiteConnect-Dart-Structs/ticker_full.dart
// To parse this JSON data, do
//
//     final tickerFull = tickerFullFromMap(jsonString);

import 'package:meta/meta.dart';
import 'dart:convert';

class TickerFull {
    TickerFull({
        @required this.averagePrice,
        @required this.buyQuantity,
        @required this.change,
        @required this.depth,
        @required this.instrumentToken,
        @required this.lastPrice,
        @required this.lastQuantity,
        @required this.lastTradeTime,
        @required this.mode,
        @required this.ohlc,
        @required this.oi,
        @required this.oiDayHigh,
        @required this.oiDayLow,
        @required this.sellQuantity,
        @required this.timestamp,
        @required this.tradable,
        @required this.volume,
    });

    final double averagePrice;
    final int buyQuantity;
    final double change;
    final Depth depth;
    final int instrumentToken;
    final double lastPrice;
    final int lastQuantity;
    final DateTime lastTradeTime;
    final String mode;
    final Ohlc ohlc;
    final int oi;
    final int oiDayHigh;
    final int oiDayLow;
    final int sellQuantity;
    final DateTime timestamp;
    final bool tradable;
    final int volume;

    TickerFull copyWith({
        double averagePrice,
        int buyQuantity,
        double change,
        Depth depth,
        int instrumentToken,
        double lastPrice,
        int lastQuantity,
        DateTime lastTradeTime,
        String mode,
        Ohlc ohlc,
        int oi,
        int oiDayHigh,
        int oiDayLow,
        int sellQuantity,
        DateTime timestamp,
        bool tradable,
        int volume,
    }) => 
        TickerFull(
            averagePrice: averagePrice ?? this.averagePrice,
            buyQuantity: buyQuantity ?? this.buyQuantity,
            change: change ?? this.change,
            depth: depth ?? this.depth,
            instrumentToken: instrumentToken ?? this.instrumentToken,
            lastPrice: lastPrice ?? this.lastPrice,
            lastQuantity: lastQuantity ?? this.lastQuantity,
            lastTradeTime: lastTradeTime ?? this.lastTradeTime,
            mode: mode ?? this.mode,
            ohlc: ohlc ?? this.ohlc,
            oi: oi ?? this.oi,
            oiDayHigh: oiDayHigh ?? this.oiDayHigh,
            oiDayLow: oiDayLow ?? this.oiDayLow,
            sellQuantity: sellQuantity ?? this.sellQuantity,
            timestamp: timestamp ?? this.timestamp,
            tradable: tradable ?? this.tradable,
            volume: volume ?? this.volume,
        );

    factory TickerFull.fromJson(String str) => TickerFull.fromMap(json.decode(str));

    String toJson() => json.encode(toMap());

    factory TickerFull.fromMap(Map<String, dynamic> json) => TickerFull(
        averagePrice: json["average_price"] == null ? null : json["average_price"].toDouble(),
        buyQuantity: json["buy_quantity"] == null ? null : json["buy_quantity"],
        change: json["change"] == null ? null : json["change"].toDouble(),
        depth: json["depth"] == null ? null : Depth.fromMap(json["depth"]),
        instrumentToken: json["instrument_token"] == null ? null : json["instrument_token"],
        lastPrice: json["last_price"] == null ? null : json["last_price"].toDouble(),
        lastQuantity: json["last_quantity"] == null ? null : json["last_quantity"],
        lastTradeTime: json["last_trade_time"] == null ? null : DateTime.parse(json["last_trade_time"]),
        mode: json["mode"] == null ? null : json["mode"],
        ohlc: json["ohlc"] == null ? null : Ohlc.fromMap(json["ohlc"]),
        oi: json["oi"] == null ? null : json["oi"],
        oiDayHigh: json["oi_day_high"] == null ? null : json["oi_day_high"],
        oiDayLow: json["oi_day_low"] == null ? null : json["oi_day_low"],
        sellQuantity: json["sell_quantity"] == null ? null : json["sell_quantity"],
        timestamp: json["timestamp"] == null ? null : DateTime.parse(json["timestamp"]),
        tradable: json["tradable"] == null ? null : json["tradable"],
        volume: json["volume"] == null ? null : json["volume"],
    );

    Map<String, dynamic> toMap() => {
        "average_price": averagePrice == null ? null : averagePrice,
        "buy_quantity": buyQuantity == null ? null : buyQuantity,
        "change": change == null ? null : change,
        "depth": depth == null ? null : depth.toMap(),
        "instrument_token": instrumentToken == null ? null : instrumentToken,
        "last_price": lastPrice == null ? null : lastPrice,
        "last_quantity": lastQuantity == null ? null : lastQuantity,
        "last_trade_time": lastTradeTime == null ? null : lastTradeTime.toIso8601String(),
        "mode": mode == null ? null : mode,
        "ohlc": ohlc == null ? null : ohlc.toMap(),
        "oi": oi == null ? null : oi,
        "oi_day_high": oiDayHigh == null ? null : oiDayHigh,
        "oi_day_low": oiDayLow == null ? null : oiDayLow,
        "sell_quantity": sellQuantity == null ? null : sellQuantity,
        "timestamp": timestamp == null ? null : timestamp.toIso8601String(),
        "tradable": tradable == null ? null : tradable,
        "volume": volume == null ? null : volume,
    };
}

class Depth {
    Depth({
        @required this.buy,
        @required this.sell,
    });

    final List<Buy> buy;
    final List<Buy> sell;

    Depth copyWith({
        List<Buy> buy,
        List<Buy> sell,
    }) => 
        Depth(
            buy: buy ?? this.buy,
            sell: sell ?? this.sell,
        );

    factory Depth.fromJson(String str) => Depth.fromMap(json.decode(str));

    String toJson() => json.encode(toMap());

    factory Depth.fromMap(Map<String, dynamic> json) => Depth(
        buy: json["buy"] == null ? null : List<Buy>.from(json["buy"].map((x) => Buy.fromMap(x))),
        sell: json["sell"] == null ? null : List<Buy>.from(json["sell"].map((x) => Buy.fromMap(x))),
    );

    Map<String, dynamic> toMap() => {
        "buy": buy == null ? null : List<dynamic>.from(buy.map((x) => x.toMap())),
        "sell": sell == null ? null : List<dynamic>.from(sell.map((x) => x.toMap())),
    };
}

class Buy {
    Buy({
        @required this.orders,
        @required this.price,
        @required this.quantity,
    });

    final int orders;
    final double price;
    final int quantity;

    Buy copyWith({
        int orders,
        double price,
        int quantity,
    }) => 
        Buy(
            orders: orders ?? this.orders,
            price: price ?? this.price,
            quantity: quantity ?? this.quantity,
        );

    factory Buy.fromJson(String str) => Buy.fromMap(json.decode(str));

    String toJson() => json.encode(toMap());

    factory Buy.fromMap(Map<String, dynamic> json) => Buy(
        orders: json["orders"] == null ? null : json["orders"],
        price: json["price"] == null ? null : json["price"].toDouble(),
        quantity: json["quantity"] == null ? null : json["quantity"],
    );

    Map<String, dynamic> toMap() => {
        "orders": orders == null ? null : orders,
        "price": price == null ? null : price,
        "quantity": quantity == null ? null : quantity,
    };
}

class Ohlc {
    Ohlc({
        @required this.close,
        @required this.high,
        @required this.low,
        @required this.open,
    });

    final double close;
    final int high;
    final double low;
    final double open;

    Ohlc copyWith({
        double close,
        int high,
        double low,
        double open,
    }) => 
        Ohlc(
            close: close ?? this.close,
            high: high ?? this.high,
            low: low ?? this.low,
            open: open ?? this.open,
        );

    factory Ohlc.fromJson(String str) => Ohlc.fromMap(json.decode(str));

    String toJson() => json.encode(toMap());

    factory Ohlc.fromMap(Map<String, dynamic> json) => Ohlc(
        close: json["close"] == null ? null : json["close"].toDouble(),
        high: json["high"] == null ? null : json["high"],
        low: json["low"] == null ? null : json["low"].toDouble(),
        open: json["open"] == null ? null : json["open"].toDouble(),
    );

    Map<String, dynamic> toMap() => {
        "close": close == null ? null : close,
        "high": high == null ? null : high,
        "low": low == null ? null : low,
        "open": open == null ? null : open,
    };
}

File: JsonSchemaAndStructs/KiteConnect-C_Plus_Plus-Structs/ticker_full.cpp
//  To parse this JSON data, first install
//
//      Boost     http://www.boost.org
//      json.hpp  https://github.com/nlohmann/json
//
//  Then include this file, and then do
//
//     TickerFull data = nlohmann::json::parse(jsonString);

#pragma once

#include "json.hpp"

#include <boost/optional.hpp>
#include <stdexcept>
#include <regex>

#ifndef NLOHMANN_OPT_HELPER
#define NLOHMANN_OPT_HELPER
namespace nlohmann {
    template <typename T>
    struct adl_serializer<std::shared_ptr<T>> {
        static void to_json(json & j, const std::shared_ptr<T> & opt) {
            if (!opt) j = nullptr; else j = *opt;
        }

        static std::shared_ptr<T> from_json(const json & j) {
            if (j.is_null()) return std::unique_ptr<T>(); else return std::unique_ptr<T>(new T(j.get<T>()));
        }
    };
}
#endif

namespace TickerFull {
    using nlohmann::json;

    inline json get_untyped(const json & j, const char * property) {
        if (j.find(property) != j.end()) {
            return j.at(property).get<json>();
        }
        return json();
    }

    inline json get_untyped(const json & j, std::string property) {
        return get_untyped(j, property.data());
    }

    template <typename T>
    inline std::shared_ptr<T> get_optional(const json & j, const char * property) {
        if (j.find(property) != j.end()) {
            return j.at(property).get<std::shared_ptr<T>>();
        }
        return std::shared_ptr<T>();
    }

    template <typename T>
    inline std::shared_ptr<T> get_optional(const json & j, std::string property) {
        return get_optional<T>(j, property.data());
    }

    class Buy {
        public:
        Buy() = default;
        virtual ~Buy() = default;

        private:
        std::shared_ptr<int64_t> orders;
        std::shared_ptr<double> price;
        std::shared_ptr<int64_t> quantity;

        public:
        std::shared_ptr<int64_t> get_orders() const { return orders; }
        void set_orders(std::shared_ptr<int64_t> value) { this->orders = value; }

        std::shared_ptr<double> get_price() const { return price; }
        void set_price(std::shared_ptr<double> value) { this->price = value; }

        std::shared_ptr<int64_t> get_quantity() const { return quantity; }
        void set_quantity(std::shared_ptr<int64_t> value) { this->quantity = value; }
    };

    class Depth {
        public:
        Depth() = default;
        virtual ~Depth() = default;

        private:
        std::shared_ptr<std::vector<Buy>> buy;
        std::shared_ptr<std::vector<Buy>> sell;

        public:
        std::shared_ptr<std::vector<Buy>> get_buy() const { return buy; }
        void set_buy(std::shared_ptr<std::vector<Buy>> value) { this->buy = value; }

        std::shared_ptr<std::vector<Buy>> get_sell() const { return sell; }
        void set_sell(std::shared_ptr<std::vector<Buy>> value) { this->sell = value; }
    };

    class Ohlc {
        public:
        Ohlc() = default;
        virtual ~Ohlc() = default;

        private:
        std::shared_ptr<double> close;
        std::shared_ptr<int64_t> high;
        std::shared_ptr<double> low;
        std::shared_ptr<double> open;

        public:
        std::shared_ptr<double> get_close() const { return close; }
        void set_close(std::shared_ptr<double> value) { this->close = value; }

        std::shared_ptr<int64_t> get_high() const { return high; }
        void set_high(std::shared_ptr<int64_t> value) { this->high = value; }

        std::shared_ptr<double> get_low() const { return low; }
        void set_low(std::shared_ptr<double> value) { this->low = value; }

        std::shared_ptr<double> get_open() const { return open; }
        void set_open(std::shared_ptr<double> value) { this->open = value; }
    };

    class TickerFull {
        public:
        TickerFull() = default;
        virtual ~TickerFull() = default;

        private:
        std::shared_ptr<double> average_price;
        std::shared_ptr<int64_t> buy_quantity;
        std::shared_ptr<double> change;
        std::shared_ptr<Depth> depth;
        std::shared_ptr<int64_t> instrument_token;
        std::shared_ptr<double> last_price;
        std::shared_ptr<int64_t> last_quantity;
        std::shared_ptr<std::string> last_trade_time;
        std::shared_ptr<std::string> mode;
        std::shared_ptr<Ohlc> ohlc;
        std::shared_ptr<int64_t> oi;
        std::shared_ptr<int64_t> oi_day_high;
        std::shared_ptr<int64_t> oi_day_low;
        std::shared_ptr<int64_t> sell_quantity;
        std::shared_ptr<std::string> timestamp;
        std::shared_ptr<bool> tradable;
        std::shared_ptr<int64_t> volume;

        public:
        std::shared_ptr<double> get_average_price() const { return average_price; }
        void set_average_price(std::shared_ptr<double> value) { this->average_price = value; }

        std::shared_ptr<int64_t> get_buy_quantity() const { return buy_quantity; }
        void set_buy_quantity(std::shared_ptr<int64_t> value) { this->buy_quantity = value; }

        std::shared_ptr<double> get_change() const { return change; }
        void set_change(std::shared_ptr<double> value) { this->change = value; }

        std::shared_ptr<Depth> get_depth() const { return depth; }
        void set_depth(std::shared_ptr<Depth> value) { this->depth = value; }

        std::shared_ptr<int64_t> get_instrument_token() const { return instrument_token; }
        void set_instrument_token(std::shared_ptr<int64_t> value) { this->instrument_token = value; }

        std::shared_ptr<double> get_last_price() const { return last_price; }
        void set_last_price(std::shared_ptr<double> value) { this->last_price = value; }

        std::shared_ptr<int64_t> get_last_quantity() const { return last_quantity; }
        void set_last_quantity(std::shared_ptr<int64_t> value) { this->last_quantity = value; }

        std::shared_ptr<std::string> get_last_trade_time() const { return last_trade_time; }
        void set_last_trade_time(std::shared_ptr<std::string> value) { this->last_trade_time = value; }

        std::shared_ptr<std::string> get_mode() const { return mode; }
        void set_mode(std::shared_ptr<std::string> value) { this->mode = value; }

        std::shared_ptr<Ohlc> get_ohlc() const { return ohlc; }
        void set_ohlc(std::shared_ptr<Ohlc> value) { this->ohlc = value; }

        std::shared_ptr<int64_t> get_oi() const { return oi; }
        void set_oi(std::shared_ptr<int64_t> value) { this->oi = value; }

        std::shared_ptr<int64_t> get_oi_day_high() const { return oi_day_high; }
        void set_oi_day_high(std::shared_ptr<int64_t> value) { this->oi_day_high = value; }

        std::shared_ptr<int64_t> get_oi_day_low() const { return oi_day_low; }
        void set_oi_day_low(std::shared_ptr<int64_t> value) { this->oi_day_low = value; }

        std::shared_ptr<int64_t> get_sell_quantity() const { return sell_quantity; }
        void set_sell_quantity(std::shared_ptr<int64_t> value) { this->sell_quantity = value; }

        std::shared_ptr<std::string> get_timestamp() const { return timestamp; }
        void set_timestamp(std::shared_ptr<std::string> value) { this->timestamp = value; }

        std::shared_ptr<bool> get_tradable() const { return tradable; }
        void set_tradable(std::shared_ptr<bool> value) { this->tradable = value; }

        std::shared_ptr<int64_t> get_volume() const { return volume; }
        void set_volume(std::shared_ptr<int64_t> value) { this->volume = value; }
    };
}

namespace nlohmann {
    void from_json(const json & j, TickerFull::Buy & x);
    void to_json(json & j, const TickerFull::Buy & x);

    void from_json(const json & j, TickerFull::Depth & x);
    void to_json(json & j, const TickerFull::Depth & x);

    void from_json(const json & j, TickerFull::Ohlc & x);
    void to_json(json & j, const TickerFull::Ohlc & x);

    void from_json(const json & j, TickerFull::TickerFull & x);
    void to_json(json & j, const TickerFull::TickerFull & x);

    inline void from_json(const json & j, TickerFull::Buy& x) {
        x.set_orders(TickerFull::get_optional<int64_t>(j, "orders"));
        x.set_price(TickerFull::get_optional<double>(j, "price"));
        x.set_quantity(TickerFull::get_optional<int64_t>(j, "quantity"));
    }

    inline void to_json(json & j, const TickerFull::Buy & x) {
        j = json::object();
        j["orders"] = x.get_orders();
        j["price"] = x.get_price();
        j["quantity"] = x.get_quantity();
    }

    inline void from_json(const json & j, TickerFull::Depth& x) {
        x.set_buy(TickerFull::get_optional<std::vector<TickerFull::Buy>>(j, "buy"));
        x.set_sell(TickerFull::get_optional<std::vector<TickerFull::Buy>>(j, "sell"));
    }

    inline void to_json(json & j, const TickerFull::Depth & x) {
        j = json::object();
        j["buy"] = x.get_buy();
        j["sell"] = x.get_sell();
    }

    inline void from_json(const json & j, TickerFull::Ohlc& x) {
        x.set_close(TickerFull::get_optional<double>(j, "close"));
        x.set_high(TickerFull::get_optional<int64_t>(j, "high"));
        x.set_low(TickerFull::get_optional<double>(j, "low"));
        x.set_open(TickerFull::get_optional<double>(j, "open"));
    }

    inline void to_json(json & j, const TickerFull::Ohlc & x) {
        j = json::object();
        j["close"] = x.get_close();
        j["high"] = x.get_high();
        j["low"] = x.get_low();
        j["open"] = x.get_open();
    }

    inline void from_json(const json & j, TickerFull::TickerFull& x) {
        x.set_average_price(TickerFull::get_optional<double>(j, "average_price"));
        x.set_buy_quantity(TickerFull::get_optional<int64_t>(j, "buy_quantity"));
        x.set_change(TickerFull::get_optional<double>(j, "change"));
        x.set_depth(TickerFull::get_optional<TickerFull::Depth>(j, "depth"));
        x.set_instrument_token(TickerFull::get_optional<int64_t>(j, "instrument_token"));
        x.set_last_price(TickerFull::get_optional<double>(j, "last_price"));
        x.set_last_quantity(TickerFull::get_optional<int64_t>(j, "last_quantity"));
        x.set_last_trade_time(TickerFull::get_optional<std::string>(j, "last_trade_time"));
        x.set_mode(TickerFull::get_optional<std::string>(j, "mode"));
        x.set_ohlc(TickerFull::get_optional<TickerFull::Ohlc>(j, "ohlc"));
        x.set_oi(TickerFull::get_optional<int64_t>(j, "oi"));
        x.set_oi_day_high(TickerFull::get_optional<int64_t>(j, "oi_day_high"));
        x.set_oi_day_low(TickerFull::get_optional<int64_t>(j, "oi_day_low"));
        x.set_sell_quantity(TickerFull::get_optional<int64_t>(j, "sell_quantity"));
        x.set_timestamp(TickerFull::get_optional<std::string>(j, "timestamp"));
        x.set_tradable(TickerFull::get_optional<bool>(j, "tradable"));
        x.set_volume(TickerFull::get_optional<int64_t>(j, "volume"));
    }

    inline void to_json(json & j, const TickerFull::TickerFull & x) {
        j = json::object();
        j["average_price"] = x.get_average_price();
        j["buy_quantity"] = x.get_buy_quantity();
        j["change"] = x.get_change();
        j["depth"] = x.get_depth();
        j["instrument_token"] = x.get_instrument_token();
        j["last_price"] = x.get_last_price();
        j["last_quantity"] = x.get_last_quantity();
        j["last_trade_time"] = x.get_last_trade_time();
        j["mode"] = x.get_mode();
        j["ohlc"] = x.get_ohlc();
        j["oi"] = x.get_oi();
        j["oi_day_high"] = x.get_oi_day_high();
        j["oi_day_low"] = x.get_oi_day_low();
        j["sell_quantity"] = x.get_sell_quantity();
        j["timestamp"] = x.get_timestamp();
        j["tradable"] = x.get_tradable();
        j["volume"] = x.get_volume();
    }
}
